lldb/source/Plugins/TypeSystem/Swift/SwiftDemangle.h

```diff
@@ -70,7 +70,77 @@ static swift::Demangle::NodePointer
 GetDemangledType(swift::Demangle::Demangler &dem, llvm::StringRef name) {
   return GetType(dem.demangleSymbol(name));
 }
+
+/// Shared ownership wrapper for swift::Demangle::NodePointer
+///
+/// Can be valid or invalid. A valid instance stores a non-null NodePointer and
+/// a shared pointer to swift::Demangle::Demangler that owns this NodePointer.
+/// An invalid instance has NodePointer equal to nullptr.
+class SharedDemangledNode {
+public:
+  SharedDemangledNode(std::shared_ptr<swift::Demangle::Demangler> demangler, llvm::StringRef name) :
+    SharedDemangledNode(demangler, demangler->demangleSymbol(name)) {}
+  
+  SharedDemangledNode(llvm::StringRef name) :
+    SharedDemangledNode(std::make_shared<swift::Demangle::Demangler>(), name) {}
+
+  SharedDemangledNode( const SharedDemangledNode& ) = default;
   
+  SharedDemangledNode& operator=( const SharedDemangledNode& ) = default;
+
+  SharedDemangledNode(SharedDemangledNode &&) = default;
+
+  SharedDemangledNode &operator=(SharedDemangledNode &&) = default;
+
+  static SharedDemangledNode Invalid() {
+    return SharedDemangledNode(nullptr, nullptr);
+  }
+
+  explicit operator bool() const { return IsValid(); }
+
+  /// Gets a raw pointer to the Node.
+  swift::Demangle::NodePointer GetRawNode() const & { return m_node; }
```

Comment:
I wonder if this function is necessary if we already implement the `->` operator.